projects/storefrontlib/src/lib/checkout/components/multi-step-checkout/delivery-mode/delivery-mode.component.ts

```diff
@@ -20,52 +20,61 @@ import { tap, takeWhile } from 'rxjs/operators';
   changeDetection: ChangeDetectionStrategy.OnPush
 })
 export class DeliveryModeComponent implements OnInit {
-  @Input()
-  selectedDeliveryMode: string;
-
-  @Output()
-  selectMode = new EventEmitter<any>();
   @Output()
-  backStep = new EventEmitter<any>();
+  goToStep = new EventEmitter<number>();
 
   supportedDeliveryModes$: Observable<DeliveryMode[]>;
-  leave = false;
+  deliveryMode: DeliveryMode;
 
   mode: FormGroup = this.fb.group({
     deliveryModeId: ['', Validators.required]
   });
 
-  constructor(private fb: FormBuilder, private service: CheckoutService) {}
+  constructor(
+    private fb: FormBuilder,
+    private checkoutService: CheckoutService,
+    private cartService: CartService
+  ) {}
 
   ngOnInit() {
-    this.supportedDeliveryModes$ = this.service
-      .getSupportedDeliveryModes()
-      .pipe(
-        takeWhile(() => !this.leave),
-        tap(supportedModes => {
-          if (Object.keys(supportedModes).length === 0) {
-            this.service.loadSupportedDeliveryModes();
-          } else {
-            if (this.selectedDeliveryMode) {
-              this.mode.controls['deliveryModeId'].setValue(
-                this.selectedDeliveryMode
-              );
-            }
-          }
-        })
-      );
+    this.checkoutService.loadSupportedDeliveryModes();
+
+    this.supportedDeliveryModes$ = combineLatest(
+      this.checkoutService.getSupportedDeliveryModes(),
+      this.checkoutService.getSelectedDeliveryMode()
+    ).pipe(
+      map(([supported, selected]: [DeliveryMode[], DeliveryMode]) => {
+        if (selected) {
+          this.mode.controls['deliveryModeId'].setValue(selected);
+          this.deliveryMode = selected;
+        }
+
+        return supported;
+      })
+    );
   }
 
   next(): void {
-    this.selectMode.emit(this.mode.value);
+    this.setDeliveryMode(this.mode.value);
+    this.goToStep.emit(3);
```

Comment:
Component should wait with redirect until setDeliveryMode method finish (when delivery mode is correctly set in our store). This solution might redirect to next step and setDeliveryMode might be still in progress. That would result in guard redirecting it back to delivery mode step.